### PR TITLE
chore(remap): add benchmark and test framework

### PR DESCRIPTION
**Pull request by JeanMertz**
This PR introduces a few convenient macros to the `remap-lang` crate to quickly add tests and benchmarks for individual functions.

It also adds a few benchmarks and tests to get started.

I'll add more in follow-up PRs, but with this merged, everyone can use the new set-up to put more of our functions under test coverage.

Built-in expressions also aren't properly tested/benched yet, these will be tackled in a follow-up PR. This also doesn't cover testing complete remap programs (or the parser specifically), again, these will be tackled separately.

### CODE REVIEW

File: lib/remap-lang/src/test_util.rs [hunk 1]
```diff
@@ -16,3 +16,79 @@ macro_rules! test_type_def {
...
+    ($($k:tt: $v:expr),+ $(,)?) => {
+        vec![$((stringify!($k), $v.into())),+]
+            .into_iter()
+            .collect::<::std::collections::HashMap<&'static str, $crate::function::Argument>>()
+    };
+}
+
+#[macro_export]
+macro_rules! bench_function {
+    ($name:tt => $func:path; $($case:ident { args: $args:expr, want: $(Ok($ok:expr))? $(Err($err:expr))? $(,)* })+) => {
```
Comment (JeanMertz) [inline lib/remap-lang/src/test_util.rs:34]: Note the `want: $(Ok($ok:expr))? $(Err($err:expr))?` part.

This is needed because the compiler can't infer the type of `T` or `E` for `Result<T, E>` if only one of the two is defined. We also can't _provide_ a value for it because we don't know the type of `T` ahead of time.

An alternative would be to allow providing the type annotation when writing the benchmark or test for the function, but this avoids having to clutter the implementation code with those annotations.

TL;DR: this ain't pretty, but it works, and makes the actual implementations nicer to read.

**Review by jszwedko — APPROVED** [review]
Looks good to me!

This is likely a personal preference, but I'm curious if the benchmarks could have been written using functions and maybe composition rather than the macros. I find the macros harder to read. Maybe I just need to spend more time with them though :smile:

**Comment by Hoverbear** [comment]
@jszwedko I also wonder if this could be done similarly with some functions. 🤔

File: Cargo.toml [hunk 2]
```diff
@@ -547,5 +547,9 @@ name = "wasm"
 harness = false
 required-features = ["transforms-wasm", "transforms-lua"]
 
+[[bench]]
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply Cargo.toml:559]: Thanks for the heads-up.

I made the necessary changes. I did move `remap` back into its own benchmark target though, as it made testing/running benchmarks much faster for me, as I didn't need to enable unneeded features, and could just run `cargo bench --bench remap`.

We can change this back in a few weeks, if you like (I'm assuming this was done to reduce compile times on the CI), once we've added more benchmarks to Remap, and we can rely more on the CI to run them, instead of running them locally.

**Comment by JeanMertz** [comment]
> I'm curious if the benchmarks could have been written using functions and maybe composition rather than the macros.

That's a fair point. I basically took what we already had for testing functions, and changed them to work for benchmarks, as it makes it easier to add both unit tests and benchmarks.

The unit tests use macro's because it allows individual tests to fail, and have more sensible names vs. having a table-driven test function. An alternative could have been smaller functions that off-load to a helper function, but I wasn't able to get a satisfactory set-up with that.

As you said, I suspect it'll just take some getting used to, but I'm happy to re-evaluate in a while and reverse course if it makes sense.
